basics/section_2.rb

```diff
@@ -0,0 +1,131 @@
+
+# # Using a method
+def said_hello(things_to_say)
+  puts things_to_say
+end
+# 
+said_hello("Hello World")
+
+#String conatenation
+first_name = "Lee"
+last_name = "RJ"
+puts first_name + " " + last_name
+
+#String interpolation
+first_name = "Lee"
+last_name = "RJ"
+puts "My first name is #{first_name} and my last name is #{last_name}"
+
+#find data type(class) with .class
+first_name = "Lee"
+last_name = "RJ"
+p first_name.class
+
+#Find methods
+first_name = "Lee"
+last_name = "RJ"
+p first_name.methods
+
+#Escape with \ (Used to escape special characters)
+name = "Ali"
+p "name is \#{name}" 
+print 'He said:\'Hi\''
+
+# `.length` sees the str length
+
+# `.reverse` reverse the str
+# `.empty?` returns a boolean to check whether the str is empty
+# `.nil?` returns a boolean to check if there is value ("".nil? returns false because empty string still has some soft of value. nil.nil? returns true)
+# Replace words with `.sub("Original string", "new string")`
+# `.gsub("OG","new") replaces all the original string to the new string stated.
+
+
+# 16 ~ 23 
+# .to_i converts str into integer
+# 10/4 will equals to 2 because ruby inteprets it as integer and drops the decimal. Have to make one of the number a float for it to show decimal. eg: 10.0/4 will equals to 2.5 or 10/4.to_f convert integers to float.
+# ruby takes it as integer divide by integer = integer. Hence, it will be an integer value. This isn’t just tied to ruby but almost all loosely typed language. Loosely typed language is a language that you do not define its type.
+# (10/4).to_f will equals to 2.0 as Ruby will evaluate the equation in the brackets first then convert the answer into float
+# You can do concat with "5" * 2 but not 2 * "5". because a string can be multiplied by two but 2 cannot multiply a string.
+# .times. For example 20.times {print "-"} will print a line of 20 dashes.
+# rand(10) generates random number between 0 and 9
+# rand generates a random number between 0 and 1
+# .to_i converrts into integer
+# if a string cant be converted into integer/float it will become 0. eg "hello".to_i = 0 "hello".to_f = 0.0
+# .odd?, .even? to check if the number is odd or even
+# comparison operators, ==, !=,>,<,>=,<=- compare classes(type)10.eql?(10.0)=false`. One is int one is float
+# define method with arguments(functions)
+# Declaration of methods has to be before using the methods.
+
+# get.chomp gets the user input, gets you a string even if input is a number.
+puts "What is ur first name?"
+first_name = gets.chomp
+puts "fname is #{first_name}"
+
+#to_i converts str into int
+puts "number to muliply by 2"
+input = gets.chomp
+puts input.to_i * 2
+
+#Analyzer program
+puts "What is your first name?"
+fname = gets.chomp
+puts "What is ur last name?"
+lname = gets.chomp
+
+fullname = fname + " " + lname
+puts "Your full name is #{fullname}"
+puts "Your reversed name is #{fullname.reverse}"
+puts "The number of characters of your name is #{fullname.length - 1}"
```

Comment:
Why fullname.length ?

Reply:
To count how many characters are there in the string which includes the spacing. Therefore the `-1`

Reply:
Ok. But doing -1 is assuming the last character is a space. What if the last character isn't a space but a valid character instead?

Reply:
I added `-1` because when I did concatenation I added `" "` in the middle so I know that I have a space to subtract. In this way we can split the string with `.split(" ")` into an array then join it again with `.join` to remove all spaces then we can calculate the characters accurately. The string will be manipulated tho.

Reply:
Don't understand. 
Let me give an example of what you just said.
```
'leerenjie'.length 
```
Is the total number of characters in your name. Does that make sense? Isn't the total number of characters in your name includes the spacing as well?

Reply:
> > > How would I know looking at this file path this file refers to which part of the course?
> > 
> > 
> > I wrote basics because the codes inside are just snippets of basic Ruby concepts. In the future, it will just be an easy source of reference for specific concepts in specific sections that are not related to the course projects in Udemy
> 
> Aku X faham Lu cakap apa. Why don't you explain the rules for the file path namings that you have set in the repo?

`basics` folder is for non-project related code. Inside that folder, I added `section_2.rb` for all the codes refereeing to section 2 of the Udemy course. In the future, if there is non-project related code in  section 3 , a `section_3.rb` file will store all the codes.